### PR TITLE
Serializer: reraise all .load errors as UnmarshalError

### PR DESCRIPTION
**Pull request by olleolleolle**
> Doesn’t _any_ error during deserialising mean the data is bad?

In order to avoid missing an error message to filter out, treat any Marshal.load error as a failed serialization, and trust Ruby's e.cause system (edit: introduced in Ruby 2.1) to provide a lineage of the error's true beginning.

This change was inspired by a comment relating to #1008.

With this rescue, it's possible to dismantle the other regular expressions.

---

This PR supersedes #1009, #1008

### CODE REVIEW

File: lib/dalli/protocol/value_serializer.rb [hunk 1]
```diff
@@ -33,42 +33,17 @@ def store(value, req_options, bitflags)
...
-
-      def filter_name_error(err)
-        raise err unless err.message.include?(NAME_ERR_STR)
-
-        raise UnmarshalError, "Unable to unmarshal value: #{err.message}"
+        if serialized
+          begin
+            serializer.load(value)
+          rescue StandardError => e
+            raise UnmarshalError, "Unable to unmarshal value: #{e.message}"
```
Comment (ioquatix) [inline lib/dalli/protocol/value_serializer.rb:42]:
```suggestion
          rescue => error
            raise UnmarshalError, "Unable to unmarshal value!"
```

Cause will already carry the message, no point repeating it.

Comment (olleolleolle) [reply lib/dalli/protocol/value_serializer.rb:42]: (The `StandardError` mention is a nod to rubocop rules.)

Comment (ioquatix) [reply lib/dalli/protocol/value_serializer.rb:42]: Actually we don't even need `=> error` if we are just raising a more general error :)

Comment (olleolleolle) [reply lib/dalli/protocol/value_serializer.rb]: 🧠 == 🧠 (I pushed new.)

File: lib/dalli/protocol/value_serializer.rb [hunk 2]
```diff
@@ -33,55 +33,17 @@ def store(value, req_options, bitflags)
...
-            begin
-              serializer.load(value)
-            rescue StandardError => e
-              raise UnmarshalError, "Unable to unmarshal value: #{e.message}"
-            end
-          else
-            # Use Dalli's existing exception filtering for deserialization when not using Marshal to serialize.
+          begin
             serializer.load(value)
+          rescue StandardError
```
Comment (ioquatix) [inline lib/dalli/protocol/value_serializer.rb:41]:
```suggestion
          rescue
```

This is identical (the default is `StandardError`).

Comment (olleolleolle) [reply lib/dalli/protocol/value_serializer.rb:41]: Yes, _and_, Rubocop rules in this project say

```

Offenses:

lib/dalli/protocol/value_serializer.rb:41:11: C: [Correctable] Style/RescueStandardError: Avoid rescuing without specifying an error class.
          rescue
          ^^^^^^
```

Comment (ioquatix) [reply lib/dalli/protocol/value_serializer.rb:41]: Haha never mind then :)

**Comment by petergoldstein** [comment]
@olleolleolle So is this an alternative to #1008 ?  

Are there any concerns on how this works with older Ruby versions?  Will they see a meaningful change in behavior?

**Comment by petergoldstein** [comment]
@olleolleolle It would helpful to understand how these different PRs relate, since they seem to touch similar code.

**Review by ioquatix — APPROVED** [review]
This is the best option by far. `cause` is supported as far back as 2.1 (https://docs.ruby-lang.org/en/2.1.0/Exception.html#method-i-cause) so I don't think depending on it for debugging failures is a problem, and in general, this PR should be much safer, since the alternative is unexpected and thus unhandled exceptions bubbling up through cache implementations (ultimately resulting in 🔥 in production).

**Comment by olleolleolle** [comment]
> @olleolleolle It would helpful to understand how these different PRs relate, since they seem to touch similar code.

Ah, thanks for asking, they're sort of the evolution of my thinking on this. 

This is the more thorough change, relying on the Ruby feature of "cause" in exception handling.

The previous ones were about "not editing too much in the project", trying to use a light touch.

**Pull request merged**